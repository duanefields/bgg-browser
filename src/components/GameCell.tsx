import { Game } from "../shared.types"
import classes from "./GameCell.module.css"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faUsers, faClock, faDice, faStar, faRankingStar } from "@fortawesome/free-solid-svg-icons"

interface GameCellProps {
  game: Game
}

const GameCell = ({ game }: GameCellProps) => {
  let playtime = `${game.minPlaytime}-${game.maxPlaytime}`
  if (game.minPlaytime === game.maxPlaytime) {
    playtime = `${game.playtime}`
  }

  let rating = `${game.averageRating} (${game.myRating})`
  if (game.myRating === 0) {
    rating = `${game.averageRating}`
  }

  return (
    <div className={classes.gridContainer}>
      <img className={classes.thumbnail} src={game.thumbnail} alt={game.name} />
      <h3 className={classes.headerArea}>{game.name}</h3>
      <div className={classes.infoArea}>
        <div>
          <FontAwesomeIcon icon={faUsers} /> {game.minPlayers}-{game.maxPlayers}
        </div>
        <div>
          <FontAwesomeIcon icon={faClock} /> {playtime} mins
        </div>
        <div>
          <FontAwesomeIcon icon={faStar} /> {rating}
        </div>
        <div>
          <FontAwesomeIcon icon={faDice} /> {game.numPlays}
        </div>
        <div>
          <FontAwesomeIcon icon={faRankingStar} /> {game.rank}
        </div>
      </div>
    </div>
  )
}

export default GameCell
