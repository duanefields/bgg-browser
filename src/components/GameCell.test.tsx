import { render, screen } from "@testing-library/react"
import { expect, it, vi } from "vitest"
import { Game } from "../shared.types"

import GameCell from "./GameCell"

const game: Game = {
  name: "Forbidden Island",
  thumbnail: "https://example.com/forbidden-island.jpg",
  minPlayers: 2,
  maxPlayers: 4,
  minPlaytime: 30,
  maxPlaytime: 45,
  playtime: 45,
  averageRating: 6.8,
  myRating: 8,
  numPlays: 10,
  rank: 100,
  objectId: 1,
  collectionId: 2,
  yearPublished: 2010,
  url: "https://example.com/ForbiddenIsland",
}

const game2: Game = {
  name: "Forbidden Island",
  thumbnail: "https://example.com/forbidden-island.jpg",
  minPlayers: 2,
  maxPlayers: 2,
  minPlaytime: 30,
  maxPlaytime: 30,
  playtime: 45,
  averageRating: 6.8,
  myRating: 0,
  numPlays: 10,
  rank: 1007,
  objectId: 1,
  collectionId: 2,
  yearPublished: 2010,
  url: "https://example.com/ForbiddenIsland",
}

it("Should match snapshot", () => {
  const cell = render(<GameCell game={game} />)
  expect(cell).toMatchSnapshot()
})

it("Should render the game name", () => {
  render(<GameCell game={game} />)
  expect(screen.getByText("Forbidden Island")).toBeVisible()
})

it("Should render the thumbnail", () => {
  render(<GameCell game={game} />)
  expect(screen.getByAltText("Forbidden Island")).toBeVisible()
})

it("Should render the player count", () => {
  render(<GameCell game={game} />)
  expect(screen.getByText("2-4")).toBeVisible()
})

it("Should render the player count if min and max are the same", () => {
  render(<GameCell game={game2} />)
  expect(screen.getByText("2")).toBeVisible()
})

it("Should render the playtime as range", () => {
  render(<GameCell game={game} />)
  expect(screen.getByText("30-45 mins")).toBeVisible()
})

it("Should render the playtime if min and max are the same", () => {
  render(<GameCell game={game2} />)
  expect(screen.getByText("45 mins")).toBeVisible()
})

it("Should render the average rating and user rating", () => {
  render(<GameCell game={game} />)
  expect(screen.getByText("6.8 (8)")).toBeVisible()
})

it("Should render the average rating only if no user rating", () => {
  render(<GameCell game={game2} />)
  expect(screen.getByText("6.8")).toBeVisible()
})

it("Should render the number of plays", () => {
  render(<GameCell game={game} />)
  expect(screen.getByText("10")).toBeVisible()
})

it("Should render the rank", () => {
  render(<GameCell game={game} />)
  expect(screen.getByText("100")).toBeVisible()
})

it("Should render the rank as a localized string", () => {
  render(<GameCell game={game2} />)
  expect(screen.getByText("1,007")).toBeVisible()
})

it("Should open the game URL in a new tab", () => {
  const winOpen = window.open
  window.open = vi.fn()
  render(<GameCell game={game} />)
  screen.getByText("Forbidden Island").click()
  expect(window.open).toHaveBeenCalledWith("https://example.com/ForbiddenIsland", "_blank")
  window.open = winOpen
})
