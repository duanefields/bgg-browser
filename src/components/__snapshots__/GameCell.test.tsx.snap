// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_gridContainer_71303f"
        data-testid="game"
      >
        <img
          alt="Forbidden Island"
          class="_thumbnail_71303f"
          height="90"
          src="https://example.com/forbidden-island.jpg"
          width="90"
        />
        <h3
          class="_headerArea_71303f"
        >
          Forbidden Island
        </h3>
        <div
          class="_infoArea_71303f"
        >
          <div>
            <svg
              aria-labelledby="svg-inline--fa-title-1"
              class="svg-inline--fa fa-users fa-fw "
              data-icon="users"
              data-prefix="fas"
              role="img"
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-1"
              >
                Player Count
              </title>
              <path
                d="M144 0a80 80 0 1 1 0 160A80 80 0 1 1 144 0zM512 0a80 80 0 1 1 0 160A80 80 0 1 1 512 0zM0 298.7C0 239.8 47.8 192 106.7 192l42.7 0c15.9 0 31 3.5 44.6 9.7c-1.3 7.2-1.9 14.7-1.9 22.3c0 38.2 16.8 72.5 43.3 96c-.2 0-.4 0-.7 0L21.3 320C9.6 320 0 310.4 0 298.7zM405.3 320c-.2 0-.4 0-.7 0c26.6-23.5 43.3-57.8 43.3-96c0-7.6-.7-15-1.9-22.3c13.6-6.3 28.7-9.7 44.6-9.7l42.7 0C592.2 192 640 239.8 640 298.7c0 11.8-9.6 21.3-21.3 21.3l-213.3 0zM224 224a96 96 0 1 1 192 0 96 96 0 1 1 -192 0zM128 485.3C128 411.7 187.7 352 261.3 352l117.3 0C452.3 352 512 411.7 512 485.3c0 14.7-11.9 26.7-26.7 26.7l-330.7 0c-14.7 0-26.7-11.9-26.7-26.7z"
                fill="currentColor"
              />
            </svg>
             
            2-4
          </div>
          <div>
            <svg
              aria-labelledby="svg-inline--fa-title-2"
              class="svg-inline--fa fa-clock fa-fw "
              data-icon="clock"
              data-prefix="fas"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-2"
              >
                Playtime
              </title>
              <path
                d="M256 0a256 256 0 1 1 0 512A256 256 0 1 1 256 0zM232 120l0 136c0 8 4 15.5 10.7 20l96 64c11 7.4 25.9 4.4 33.3-6.7s4.4-25.9-6.7-33.3L280 243.2 280 120c0-13.3-10.7-24-24-24s-24 10.7-24 24z"
                fill="currentColor"
              />
            </svg>
             
            30-45
             mins
          </div>
          <div>
            <svg
              aria-labelledby="svg-inline--fa-title-3"
              class="svg-inline--fa fa-star fa-fw "
              data-icon="star"
              data-prefix="fas"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-3"
              >
                BGG Rating (Your Rating)
              </title>
              <path
                d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                fill="currentColor"
              />
            </svg>
             
            6.8 (8)
          </div>
          <div>
            <svg
              aria-labelledby="svg-inline--fa-title-4"
              class="svg-inline--fa fa-dice fa-fw "
              data-icon="dice"
              data-prefix="fas"
              role="img"
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-4"
              >
                Number of Plays
              </title>
              <path
                d="M274.9 34.3c-28.1-28.1-73.7-28.1-101.8 0L34.3 173.1c-28.1 28.1-28.1 73.7 0 101.8L173.1 413.7c28.1 28.1 73.7 28.1 101.8 0L413.7 274.9c28.1-28.1 28.1-73.7 0-101.8L274.9 34.3zM200 224a24 24 0 1 1 48 0 24 24 0 1 1 -48 0zM96 200a24 24 0 1 1 0 48 24 24 0 1 1 0-48zM224 376a24 24 0 1 1 0-48 24 24 0 1 1 0 48zM352 200a24 24 0 1 1 0 48 24 24 0 1 1 0-48zM224 120a24 24 0 1 1 0-48 24 24 0 1 1 0 48zm96 328c0 35.3 28.7 64 64 64l192 0c35.3 0 64-28.7 64-64l0-192c0-35.3-28.7-64-64-64l-114.3 0c11.6 36 3.1 77-25.4 105.5L320 413.8l0 34.2zM480 328a24 24 0 1 1 0 48 24 24 0 1 1 0-48z"
                fill="currentColor"
              />
            </svg>
             
            1,001
          </div>
          <div>
            <svg
              aria-labelledby="svg-inline--fa-title-5"
              class="svg-inline--fa fa-ranking-star fa-fw "
              data-icon="ranking-star"
              data-prefix="fas"
              role="img"
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-5"
              >
                BGG Rank
              </title>
              <path
                d="M353.8 54.1L330.2 6.3c-3.9-8.3-16.1-8.6-20.4 0L286.2 54.1l-52.3 7.5c-9.3 1.4-13.3 12.9-6.4 19.8l38 37-9 52.1c-1.4 9.3 8.2 16.5 16.8 12.2l46.9-24.8 46.6 24.4c8.6 4.3 18.3-2.9 16.8-12.2l-9-52.1 38-36.6c6.8-6.8 2.9-18.3-6.4-19.8l-52.3-7.5zM256 256c-17.7 0-32 14.3-32 32l0 192c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-192c0-17.7-14.3-32-32-32l-128 0zM32 320c-17.7 0-32 14.3-32 32L0 480c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-128c0-17.7-14.3-32-32-32L32 320zm416 96l0 64c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-64c0-17.7-14.3-32-32-32l-128 0c-17.7 0-32 14.3-32 32z"
                fill="currentColor"
              />
            </svg>
             
            1,007
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_gridContainer_71303f"
      data-testid="game"
    >
      <img
        alt="Forbidden Island"
        class="_thumbnail_71303f"
        height="90"
        src="https://example.com/forbidden-island.jpg"
        width="90"
      />
      <h3
        class="_headerArea_71303f"
      >
        Forbidden Island
      </h3>
      <div
        class="_infoArea_71303f"
      >
        <div>
          <svg
            aria-labelledby="svg-inline--fa-title-1"
            class="svg-inline--fa fa-users fa-fw "
            data-icon="users"
            data-prefix="fas"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="svg-inline--fa-title-1"
            >
              Player Count
            </title>
            <path
              d="M144 0a80 80 0 1 1 0 160A80 80 0 1 1 144 0zM512 0a80 80 0 1 1 0 160A80 80 0 1 1 512 0zM0 298.7C0 239.8 47.8 192 106.7 192l42.7 0c15.9 0 31 3.5 44.6 9.7c-1.3 7.2-1.9 14.7-1.9 22.3c0 38.2 16.8 72.5 43.3 96c-.2 0-.4 0-.7 0L21.3 320C9.6 320 0 310.4 0 298.7zM405.3 320c-.2 0-.4 0-.7 0c26.6-23.5 43.3-57.8 43.3-96c0-7.6-.7-15-1.9-22.3c13.6-6.3 28.7-9.7 44.6-9.7l42.7 0C592.2 192 640 239.8 640 298.7c0 11.8-9.6 21.3-21.3 21.3l-213.3 0zM224 224a96 96 0 1 1 192 0 96 96 0 1 1 -192 0zM128 485.3C128 411.7 187.7 352 261.3 352l117.3 0C452.3 352 512 411.7 512 485.3c0 14.7-11.9 26.7-26.7 26.7l-330.7 0c-14.7 0-26.7-11.9-26.7-26.7z"
              fill="currentColor"
            />
          </svg>
           
          2-4
        </div>
        <div>
          <svg
            aria-labelledby="svg-inline--fa-title-2"
            class="svg-inline--fa fa-clock fa-fw "
            data-icon="clock"
            data-prefix="fas"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="svg-inline--fa-title-2"
            >
              Playtime
            </title>
            <path
              d="M256 0a256 256 0 1 1 0 512A256 256 0 1 1 256 0zM232 120l0 136c0 8 4 15.5 10.7 20l96 64c11 7.4 25.9 4.4 33.3-6.7s4.4-25.9-6.7-33.3L280 243.2 280 120c0-13.3-10.7-24-24-24s-24 10.7-24 24z"
              fill="currentColor"
            />
          </svg>
           
          30-45
           mins
        </div>
        <div>
          <svg
            aria-labelledby="svg-inline--fa-title-3"
            class="svg-inline--fa fa-star fa-fw "
            data-icon="star"
            data-prefix="fas"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="svg-inline--fa-title-3"
            >
              BGG Rating (Your Rating)
            </title>
            <path
              d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
              fill="currentColor"
            />
          </svg>
           
          6.8 (8)
        </div>
        <div>
          <svg
            aria-labelledby="svg-inline--fa-title-4"
            class="svg-inline--fa fa-dice fa-fw "
            data-icon="dice"
            data-prefix="fas"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="svg-inline--fa-title-4"
            >
              Number of Plays
            </title>
            <path
              d="M274.9 34.3c-28.1-28.1-73.7-28.1-101.8 0L34.3 173.1c-28.1 28.1-28.1 73.7 0 101.8L173.1 413.7c28.1 28.1 73.7 28.1 101.8 0L413.7 274.9c28.1-28.1 28.1-73.7 0-101.8L274.9 34.3zM200 224a24 24 0 1 1 48 0 24 24 0 1 1 -48 0zM96 200a24 24 0 1 1 0 48 24 24 0 1 1 0-48zM224 376a24 24 0 1 1 0-48 24 24 0 1 1 0 48zM352 200a24 24 0 1 1 0 48 24 24 0 1 1 0-48zM224 120a24 24 0 1 1 0-48 24 24 0 1 1 0 48zm96 328c0 35.3 28.7 64 64 64l192 0c35.3 0 64-28.7 64-64l0-192c0-35.3-28.7-64-64-64l-114.3 0c11.6 36 3.1 77-25.4 105.5L320 413.8l0 34.2zM480 328a24 24 0 1 1 0 48 24 24 0 1 1 0-48z"
              fill="currentColor"
            />
          </svg>
           
          1,001
        </div>
        <div>
          <svg
            aria-labelledby="svg-inline--fa-title-5"
            class="svg-inline--fa fa-ranking-star fa-fw "
            data-icon="ranking-star"
            data-prefix="fas"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="svg-inline--fa-title-5"
            >
              BGG Rank
            </title>
            <path
              d="M353.8 54.1L330.2 6.3c-3.9-8.3-16.1-8.6-20.4 0L286.2 54.1l-52.3 7.5c-9.3 1.4-13.3 12.9-6.4 19.8l38 37-9 52.1c-1.4 9.3 8.2 16.5 16.8 12.2l46.9-24.8 46.6 24.4c8.6 4.3 18.3-2.9 16.8-12.2l-9-52.1 38-36.6c6.8-6.8 2.9-18.3-6.4-19.8l-52.3-7.5zM256 256c-17.7 0-32 14.3-32 32l0 192c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-192c0-17.7-14.3-32-32-32l-128 0zM32 320c-17.7 0-32 14.3-32 32L0 480c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-128c0-17.7-14.3-32-32-32L32 320zm416 96l0 64c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-64c0-17.7-14.3-32-32-32l-128 0c-17.7 0-32 14.3-32 32z"
              fill="currentColor"
            />
          </svg>
           
          1,007
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
